associations/serializers/association.py

```diff
@@ -1,110 +1,108 @@
+                                      if permission_name != 'administration') +
+                                ('is_archived', 'role', 'rank'))
 
-            message = self.error_messages['empty']
-            raise ValidationError({
-                'test': [message]
-            }, code='empty')
+            for field in forbidden_fields:
+                if field in self.validated_data:
+                    self.validated_data.pop(field)
 
-        ret = []
-        errors = []
+        return super(WriteRoleSerializer, self).save(**kwargs)
 
-        for item in data:
-            try:
-                # Code that was inserted
-                self.child.instance = self.instance.get(id=item['id']) if self.instance else None
-                self.child.initial_data = item
-                # Until here
-                validated = self.child.run_validation(item)
-            except ValidationError as exc:
-                errors.append(exc.detail)
-            else:
-                ret.append(validated)
-                errors.append({})
+    def update(self, instance, validated_data):
+        # Prevent the changes to association and user fields when updating.
+        if 'association' in validated_data:
+            validated_data.pop('association')
 
-        if any(errors):
-            raise ValidationError(errors)
+        if 'user' in validated_data:
+            validated_data.pop('user')
 
-        return ret
+        return super(WriteRoleSerializer, self).update(instance, validated_data)
 
+    def to_representation(self, instance):
+        res = super(WriteRoleSerializer, self).to_representation(instance)
 
-class AdaptedBulkListSerializer(AdaptedBulkListSerializerMixin, BulkListSerializer):
-    pass
+        for permission_name in Role.PERMISSION_NAMES:
+            res[permission_name] = getattr(instance, permission_name)
 
+        return res
 
-class RoleSerializer(BulkSerializerMixin, serializers.ModelSerializer):
     class Meta:
         model = Role
-        fields = "__all__"
-        list_serializer_class = AdaptedBulkListSerializer
+        read_only_fields = ('id',)
+        fields = (read_only_fields +
+                  ('association', 'user', 'role', 'rank', 'is_archived') +
+                  tuple(f'{permission_name}_permission' for permission_name in Role.PERMISSION_NAMES))
+
+
+class RoleSerializer(serializers.ModelSerializer):
+    user = UserShortSerializer(read_only=True)
 
     def to_representation(self, instance):
-        response = super().to_representation(instance)
-        response['user'] = UserShortSerializer(instance.user).data
-        return response
+        res = super(RoleSerializer, self).to_representation(instance)
+
+        for permission_name in Role.PERMISSION_NAMES:
+            res[permission_name] = getattr(instance, permission_name)
+
+        return res
+
+    class Meta:
+        model = Role
+        fields = '__all__'
```

Comment:
It's possible to change the `association` with this serializer. Perhaps it's better to change this to removing `to_representation` and declaring all fields as `read_only_field`(https://stackoverflow.com/questions/38319856/django-rest-framework-read-only-model-serializer)

Reply:
It is indeed.
Problem is, the fields in to_representation are actually properties, and DRF doesn't seem to consider them as regular fields...
But you are right, this Serializer should definitely be read-only, as one may be using it for writing data without paying attention (actually happened to me...) Maybe we could override save to raise an exception if it is used?